backend/executor.go

```diff
@@ -321,10 +339,27 @@ func (g *grpcExecutor) StartInstance(ctx context.Context, req *protos.CreateInst
 // TerminateInstance implements protos.TaskHubSidecarServiceServer
 func (g *grpcExecutor) TerminateInstance(ctx context.Context, req *protos.TerminateRequest) (*protos.TerminateResponse, error) {
 	e := helpers.NewExecutionTerminatedEvent(req.Output, req.Recursive)
-	if err := g.backend.AddNewOrchestrationEvent(ctx, api.InstanceID(req.InstanceId), e); err != nil {
-		return nil, err
+	id := api.InstanceID(req.InstanceId)
+	// Parent instance needs to be terminated first
+	if err := g.backend.AddNewOrchestrationEvent(ctx, id, e); err != nil {
+		return nil, fmt.Errorf("failed to terminate orchestration instance %s: %w", id, err)
+	}
+	if req.Recursive {
+		subOrchestrationInstances, err := GetSubOrchestrationInstances(ctx, g.backend, id, false)
```

Comment:
Same comment as client.go about this approach being prone to race conditions.

Reply:
Done